### PR TITLE
Optimize GetMemberType() extension; TypeHelper

### PR DESCRIPTION
**Pull request by SergeiPavlov**
* Memoize results of `Type` -> `MemberType` mapping in  `GetMemberType()`
* Avoid multiple calling `StripMarkers()` for the same expression

### CODE REVIEW

File: Orm/Xtensive.Orm/Orm/Linq/ExpressionExtensions.cs [hunk 1]
```diff
@@ -152,63 +151,31 @@ public static bool IsMarker(this Expression e)
...
         markerType = marker.MarkerType;
+        return true;
       }
-
-      return result;
+      markerType = MarkerType.None;
+      return false;
     }
 
     public static MemberType GetMemberType(this Expression e)
```
Comment (botinko) [inline Orm/Xtensive.Orm/Orm/Linq/ExpressionExtensions.cs:162]: Not sure, that new version is faster. Need a benchmark to proof.

Comment (botinko) [reply Orm/Xtensive.Orm/Orm/Linq/ExpressionExtensions.cs:162]: I'd for optimizing by using MetadataToken as a key and use memoize

File: Orm/Xtensive.Orm/Reflection/TypeHelper.cs [hunk 2]
```diff
@@ -992,8 +995,11 @@ public static Type GetGenericType(this Type type, Type openGenericBaseType)
     /// where type parameters are bound in case it is implemented by the <paramref name="type"/>;
     /// otherwise, <see langword="null"/>.
     /// </returns>
-    public static Type GetGenericInterface(this Type type, Type openGenericInterface)
-    {
+    public static Type GetGenericInterface(this Type type, Type openGenericInterface) =>
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Reflection/TypeHelper.cs]: rolled back it

File: Orm/Xtensive.Orm/Orm/Linq/ExpressionExtensions.cs (hunk 1 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Linq/ExpressionExtensions.cs:162]: Using MetadataToken for comparison will change semantics for TypeDelegator.

Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Linq/ExpressionExtensions.cs:164]: Changed dictionary key to tuple `(type.MetadataToken, type.Module)`

**Pull request merged**